docs/source/ser_deser.rst

```diff
@@ -0,0 +1,78 @@
+
+High level serialization and deserialization flow
+=================================================
+
+```python
+import copy
+import tempfile
+import torch
+from torchao.quantization.quant_api import (
+    quantize_,
+    int4_weight_only,
+)
+
+class ToyLinearModel(torch.nn.Module):
+    def __init__(self, m=64, n=32, k=64):
+        super().__init__()
+        self.linear1 = torch.nn.Linear(m, n, bias=False)
+        self.linear2 = torch.nn.Linear(n, k, bias=False)
+
+    def example_inputs(self, batch_size=1, dtype=torch.float32, device="cpu"):
+        return (torch.randn(batch_size, self.linear1.in_features, dtype=dtype, device=device),)
+
+    def forward(self, x):
+        x = self.linear1(x)
+        x = self.linear2(x)
+        return x
+
+dtype = torch.bfloat16
+m = ToyLinearModel(1024, 1024, 1024).eval().to(dtype).to("cuda")
+example_inputs = m.example_inputs(dtype=dtype, device="cuda")
+quantize_(m, int4_weight_only())
+ref = m(*example_inputs)
+with tempfile.NamedTemporaryFile() as f:
+    torch.save(m.state_dict(), f)
+    f.seek(0)
+    state_dict = torch.load(f)
+
+with torch.device("meta"):
+    m_loaded = ToyLinearModel(1024, 1024, 1024).eval().to(dtype)
+    
+m_loaded.load_state_dict(state_dict, assign=True)
+res = m_loaded(*example_inputs)
+assert torch.equal(res, ref)
+```
+
+What happens when serializing an optimized model?
+=================================================
+To serilize an optimized model, we just need to call `torch.save(m.state_dict(), f)`, because in torchao, we use tensor subclass to represent different dtypes or support different optimization techniques like quantization and sparsity. So after optimization, the only thing that is updated is the weight Tensor and the model structure is not changed at all. For example:
+
+original floating point model `state_dict`:
+```
+{"linear1.weight": float_weight1, "linear2.weight": float_weight2}
+```
+
+quantized model `state_dict`:
+```
+{"linear1.weight": quantized_weight1, "linear2.weight": quantized_weight2, ...}
+```
+
+The size of the quantized model is typically going to be smaller to the original floating point model, but it also depends on the specific techinque and implementation you are using.
+
+What happens when deserializing an optimized model?
+===================================================
+To deserialize an optimized model, we can initialize the floating point model in meta device and then load the optimized `state_dict` with `assign=True`:
+
+```
+with torch.device("meta"):
+    m_loaded = ToyLinearModel(1024, 1024, 1024).eval().to(dtype)
+    
+m_loaded.load_state_dict(state_dict, assign=True)
+```
+
+What happens in `m_loaded.load_state_dict(state_dict, assign=True)` is that the corresponding weights (e.g. m_loaded.linear1.weight) are updated with the Tensors in `state_dict`, which is an optimized tensor subclass instance (e.g. int4 quantized tensor). No dependency on torchao is needed for this to work.
```

Comment:
what's not clear though is after loading the model in this manner are the weights still in a lower bit dtype?

Reply:
the weights will be updated to lower bit dtype tensor (like int4 AffineQuantizedTensor), are you talking about what is `model.linear1.weight.dtype`?

Reply:
Basically how do you verify that the load happened correctly and that you've now loaded a quantized model - 2 ideas are 
1. You might either run the model and check that its faster or
2. Printing the names and dtypes of all layers in `m_loaded`

Reply:
we can print the type of weights I think, I can add this